ptvsd/wrapper.py

```diff
@@ -737,19 +739,43 @@ def on_variables(self, request, args):
         except AttributeError:
             xvars = []
 
-        variables = []
+        variables = [] # variables that do not begin with underscores
+        single_underscore = [] # variables that begin with underscores
+        double_underscore = [] # variables that begin with two underscores
+        dunder = [] # variables that begin and end with double underscores
         for xvar in xvars:
+            var_name = unquote(xvar['name'])
+            var_type = unquote(xvar['type'])
+            var_value = unquote(xvar['value'])
             var = {
-                'name': unquote(xvar['name']),
-                'type': unquote(xvar['type']),
-                'value': unquote(xvar['value']),
+                'name': var_name,
+                'type': var_type,
+                'value': var_value,
             }
             if bool(xvar['isContainer']):
-                pyd_child = pyd_var + (var['name'],)
+                pyd_child = pyd_var + (var_name,)
                 var['variablesReference'] = self.var_map.to_vscode(
                     pyd_child, autogen=True)
-            variables.append(var)
 
+            if var_name.startswith('__'):
+                if var_name.endswith('__'):
+                    dunder.append(var)
+                else:
+                    double_underscore.append(var)
+            elif var_name.startswith('_'):
+                single_underscore.append(var)
+            else:
+                variables.append(var)
+
+        def get_sort_key(o):
+            return o['name']
+
+        variables.sort(key=get_sort_key)
+        single_underscore.sort(key=get_sort_key)
+        double_underscore.sort(key=get_sort_key)
+        dunder.sort(key=get_sort_key)
+
+        variables = variables + single_underscore + double_underscore + dunder
```

Comment:
I'd move all of this into a separate class to track and sort the variables.
If tomorrow we need to sort private variables or member variables, we need to modify this method again, when logic of sorting the list should be outside.

Reply:
This stuff or sorting and tracking variables should be in its own class, something that can be easily unit tested in isolation.

Reply:
I think we'll need unit tests for this sorting rule.

Reply:
I will move this to a class before merging.

Reply:
:+1: